app/src/main/java/com/example/android/betterdiceroll/GameController.java

```diff
@@ -0,0 +1,311 @@
+package com.example.android.betterdiceroll;
+
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+
+public class GameController
+        implements Bundleable {
+    private Random random = new Random();
+
+    private int currentScore = 0;
+    private int rollCount = 3;
+    private ArrayList<Dice> dices = new ArrayList<>();
+
+    private boolean gameEnded;
+    private boolean rollDisabled;
+
+    {
+        for(int i = 0; i < 5; i++) {
+            Dice dice = new Dice(i);
+            dices.add(i, dice);
+        }
+    }
+
+    public int getCurrentScore() {
+        return currentScore;
+    }
+
+    public int getRollCount() {
+        return rollCount;
+    }
+
+    public List<Dice> getDices() {
+        return Collections.unmodifiableList(dices);
+    }
+
+    public boolean isGameEnded() {
+        return gameEnded;
+    }
+
+    @NonNull
+    @Override
+    public Bundle toBundle() {
+        Bundle bundle = new Bundle();
+        bundle.putInt("currentScore", currentScore);
+        bundle.putInt("rollCount", rollCount);
+        bundle.putSerializable("dices", dices);
+        bundle.putBoolean("gameEnded", gameEnded);
+        bundle.putBoolean("rollDisabled", rollDisabled);
+        return bundle;
+    }
+
+    @Override
+    public void fromBundle(@Nullable Bundle bundle) {
+        if(bundle != null) {
+            currentScore = bundle.getInt("currentScore");
+            rollCount = bundle.getInt("rollCount");
+            //noinspection unchecked
+            dices = ((ArrayList<Dice>) bundle.getSerializable("dices"));
+            gameEnded = bundle.getBoolean("gameEnded");
+            rollDisabled = bundle.getBoolean("rollDisabled");
+        }
+    }
+
+    public void roll() {
+        if(rollCount > 0) {
+            rollDiceIfActive();
+            rollCount--;
+            if(gameUpdateListener != null) {
+                gameUpdateListener.onRollCountChanged(rollCount);
+            }
+        }
+
+        if(rollCount == 0) {
+            rollDisabled = true;
+            if(gameUpdateListener != null) {
+                gameUpdateListener.onRollDisabled();
+                boolean areThereDiceToHold = false;
```

Comment:
Whoops, this line should be outside of the `if`.

Reply:
it's also an incomplete duplication of the game finalization logic. What was I thinking :smile: